### DISCUSSION TITLE
Music notation support

### DISCUSSION THREAD

**Discussion by batonac** [praisenter/praisenter#38] [Ideas]
How difficult would it be to implement music notation support on song slides?

A good illustration of this can be found here: https://my.hymnary.org/flexpresent-preview/

The underlying library powering that demonstration is: https://gitlab.ccel.org/hymnary/dynamic-presentation

In that case, the input format is musicXML.

**Comment by wnbittle** [comment]
Just so I have a clear understanding of the request - you would like to create a slide template that would show song lyrics along side music notation like this:

![image](https://github.com/praisenter/praisenter/assets/11467997/8eaf943d-cfe4-4ceb-b385-8731fcb521dc)

If so, here are few challenges that I'm seeing:

- The Verovio library (which is used to do the music notation rendering) has Java bindings, but it's not released on Maven so it would need to be packaged as a CLI or manually referenced.  Both of which are less than ideal.
- The output of Verovio is SVG which is great, but the rendering engine that Praisenter is build on (Java FX) doesn't have full SVG support.  It may work correctly, half-way, or not at all.  A proof of concept would be needed.
- For song lyric entry, the only realistic way would be to import a musicXML formatted file.  Trying to replicate an editor for this complex of a file would be very challenging (not impossible though).
- The musicXML file has the song split into parts (the one example I opened seemed to be split by treble & base clef) and parts split into measures.  The measures then contain the notes and what lyric they are a part of.  I didn't have time to research more, but there would need to be a way to split this into the format Praisenter is looking for (by lyric).

As to difficulty - it's a challenging request for sure.

---
To get a proof of concept working (for me):

1. Get the Verovio library callable from Praisenter (Java) in a supportable way
2. Run a few musicXML files through Verovio to get the SVG and attempt to display that in Praisenter (if all is well, then continue, else we'd need to investigation an SVG to PNG conversion strategy)
3. Determine a way to keep the existing song format, but allow additional formats like this to live side-by-side
4. Determine what an "Editor" would look like for this
5. Determine how to handle slide templates with this content

**Comment by batonac** [comment]
You got the vision, for sure! This is a big deal for congregations that sing hymns with vocal harmony.

I did just a bit of digging into these hurdles. I'm not familiar with Java development, but I noticed that [the Guido project](https://github.com/grame-cncm/guidolib) provides both a plain text music notation format, which could presumably fill the gap left by the lack of a built-in graphical editor, as it would be much easier to edit by hand compared to MusicXML. The music notation itself would be tedious to edit for most users, but it seems that lyrical edits should be quite doable.

[libmusicxml](https://github.com/grame-cncm/libmusicxml) provides the ability to convert from MusicXML to Guido syntax, so that could presumably provide a nice workflow for song import, allowing songs to be typeset in any software that exports to MusicXML (MuseScore or equivalent) initially and then converted to Guido on import into Praisenter.

The Guido project has native Java bindings, but I'm not sure how to verify if it provides better rendering options with support for Maven and Java FX.

**Reply by batonac** [reply]
It appears that the applicable [Guido rendering code for Java is likely here](https://github.com/grame-cncm/guidolib/blob/dev/lang/java/guidoviewer.java).

**Reply by wnbittle** [reply]
Nice, I'll take a look at those.  Guido appears to use AWT/Swing and there is a Java FX bridge to Swing.

Another workflow might be to generate the rendering ahead of time and save it as an image.  I'd rather have a way to render directly though since there would be higher quality scaling options.

**Comment by reckel-jm** [comment]
Hey, I am the main developer and maintainer of [Cantara](https://github.com/reckel-jm/cantara), a similar but much simplier open source project for displaying songs in churches. We are also planning to implement score presenting in the foreseeable future. In our case, we decided to create our own song format and use Lilypond as the music renderer. The song files would basically look like this:

Amazing Grace.cssf
```
#title: Amazing Grace
#author: John Newton
#language: en,de
#chord: f major
#time: 3/4
#partial: 4

#stanza1.mainscore
c4 | f2 a8( f) | a2 g4 | f2 d4 | c2 \breathe c4 | f2 a8( f) | a2 g4 | c2.~ | c2 \breathe a4 |
  c2 a8( f ) | a2 g4 | f2 d4 | c2 \breathe c4 | f2 a8( f ) | a2 g4 | f2. 

#stanza1.lyrics.en
A -- ma -- zing grace, How sweet the sound
That saved a wretch like me.
I once was lost, but now I'm found,
Was blind, but now I see.

#stanza1.lyrics.de
O kommt und staunt und be -- tet an
den Her -- ren Je -- sus Christ
Er hat so Gro -- ßes mir ge -- tan,
das nicht zu sa -- gen ist.

#stanza1.chords
\skip 4 | f2. | f:7 | b | f | d:m | f | c | c | f | f:7 | b | f | d:m | b2 c4 | f2.
```
This can be translated to a song presentation slide with lyrics but without scores, a lilypond file for the musicians and a presentation file with scores.

This just as an idea. I wish you good luck and God's blessing for the development!

**Reply by batonac** [reply]
@reckel-jm this looks very interesting, and kudos for a very user-friendly syntax! Are you planning to create any import/migration path from MusicXML?

**Reply by reckel-jm** [reply]
No, this is not planned at the moment. The reason for this is that MusicXML is basically a score description syntax wich describes scores in their order and structure but can not describe songs with its parts. There is a difference between them as the idea is to generate scores depending on the song AND the desired output.
Let me clarify that with an example. Let's say we have the song Amazing Grace as mentioned above with three stanzas. If you would like to print the scores for the musicians you would assign the three stanzas **_parallel_** (under) each other. 
However, if you want to make a presentation, you would need to assign the stanzas **_next to_** (behind) each other. We need a file format which handles the structure **and** the content of songs. This can be used to generate any output of any choice later. MusicXML is not a handy format for hand typing. The Lilypond syntax is IMHO much better in terms of usability, that's why we at Cantara have no plans to use MusicXML at the moment.

**Reply by wnbittle** [reply]
@reckel-jm  for your solution could you answer a few questions about what I understand your steps to be?

1. Convert your format (one stanza or the whole thing?) to lilypond format
2. Run lilypond (are you planning to package it with your software?) to produce something renderable (SVG, PNG or ?)
3. Display the renderable on a presentation

At what point in the lifecycle are you planning to do the conversion - at render time or at save time or?

**Reply by reckel-jm** [reply]
Yes, your steps are correct. Lilypond supports several output formats (PDF, SVG, PNG). All of them can be either printed on a page, or cropped. This should happen before the presentation starts and can then be inserted into the presentation.

1. Lilypond can be configured to export each line in an own file. So I can just give the whole sing as input and expect several output files.

2. I have not decided on that yet. Most probably I will not include it by myself, but it will be an optional dependency on the program.
